File: sdk/aip/golang/identity.go
package aip

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"time"
)

type AIPIdentity struct {
	DID        string
	PublicKey  ed25519.PublicKey
	privateKey ed25519.PrivateKey
}

func NewAIPIdentity() (*AIPIdentity, error) {
	pub, priv, err := ed25519.GenerateKey(nil)
	if err != nil { return nil, fmt.Errorf("generate key: %w", err) }
	did, _ := PublicKeyToDIDKey(pub)
	return &AIPIdentity{DID: did, PublicKey: pub, privateKey: priv}, nil
}

func AIPIdentityFromAPIKey(apiKey string) (*AIPIdentity, error) {
	seed := sha256.Sum256([]byte(apiKey))
	priv := ed25519.NewKeyFromSeed(seed[:])
	pub := priv.Public().(ed25519.PublicKey)
	did, _ := PublicKeyToDIDKey(pub)
	return &AIPIdentity{DID: did, PublicKey: pub, privateKey: priv}, nil
}

func AIPIdentityFromPrivateKey(privB64 string) (*AIPIdentity, error) {
	seed, err := base64.StdEncoding.DecodeString(privB64)
	if err != nil { return nil, fmt.Errorf("decode: %w", err) }
	if len(seed) != ed25519.SeedSize { return nil, errors.New("invalid key length") }
	priv := ed25519.NewKeyFromSeed(seed)
	pub := priv.Public().(ed25519.PublicKey)
	did, _ := PublicKeyToDIDKey(pub)
	return &AIPIdentity{DID: did, PublicKey: pub, privateKey: priv}, nil
}

func (id *AIPIdentity) PublicKeyBase64() string { return base64.StdEncoding.EncodeToString(id.PublicKey) }
func (id *AIPIdentity) Sign(data []byte) string { return base64.StdEncoding.EncodeToString(ed25519.Sign(id.privateKey, data)) }
func (id *AIPIdentity) ExportPrivateKey() string { return base64.StdEncoding.EncodeToString(id.privateKey.Seed()) }

func AIPVerify(data []byte, sigB64 string, signerDID string) bool {
	pub, err := DIDKeyToPublicKey(signerDID)
	if err != nil { return false }
	sig, err := base64.StdEncoding.DecodeString(sigB64)
	if err != nil { return false }
	return ed25519.Verify(pub, data, sig)
}

func (id *AIPIdentity) GetDIDDocument() map[string]interface{} {
	keyID := id.DID + "#keys-1"
	now := time.Now().UTC().Format(time.RFC3339)
	return map[string]interface{}{
		"@context": []string{"https://www.w3.org/ns/did/v1", "https://w3id.org/security/suites/ed25519-2020/v1"},
		"id": id.DID, "controller": id.DID,
		"verificationMethod": []map[string]string{{"id": keyID, "type": "Ed25519VerificationKey2020", "controller": id.DID, "publicKeyMultibase": id.DID[8:]}},
		"authentication": []string{keyID}, "assertionMethod": []string{keyID},
		"capabilityDelegation": []string{keyID}, "capabilityInvocation": []string{keyID},
		"created": now, "updated": now,
	}
}

File: sdk/aip/typescript/src/identity.ts
/**
 * AIPIdentity — standalone DID identity class (no platform dependency)
 */

import { publicKeyToDIDKey, didKeyToPublicKey } from './did';

export interface DIDDocument {
  '@context': string[];
  id: string;
  controller: string;
  verificationMethod: { id: string; type: string; controller: string; publicKeyMultibase: string }[];
  authentication: string[];
  assertionMethod: string[];
  capabilityDelegation: string[];
  capabilityInvocation: string[];
  service?: { id: string; type: string; serviceEndpoint: string }[];
  'aip:capabilities'?: string[];
  created: string;
  updated: string;
}

export interface SignedPayload {
  senderDid: string;
  signature: string;
  payload: Uint8Array;
}

let _ed25519: any = null;

async function getEd25519() {
  if (_ed25519) return _ed25519;
  const mod = await import('@noble/curves/ed25519.js');
  _ed25519 = mod.ed25519;
  return _ed25519;
}

export class AIPIdentity {
  readonly did: string;
  readonly publicKey: Uint8Array;
  private readonly privateKey: Uint8Array;

  private constructor(priv: Uint8Array, pub: Uint8Array, did: string) {
    this.privateKey = priv;
    this.publicKey = pub;
    this.did = did;
  }

  static async create(): Promise<AIPIdentity> {
    const ed = await getEd25519();
    const priv = ed.utils.randomSecretKey();
    const pub = ed.getPublicKey(priv);
    return new AIPIdentity(priv, pub, publicKeyToDIDKey(pub));
  }

  static async fromApiKey(apiKey: string): Promise<AIPIdentity> {
    const ed = await getEd25519();
    const keyBytes = new TextEncoder().encode(apiKey);
    const hashBuffer = await crypto.subtle.digest('SHA-256', keyBytes);
    const seed = new Uint8Array(hashBuffer);
    const pub = ed.getPublicKey(seed);
    return new AIPIdentity(seed, pub, publicKeyToDIDKey(pub));
  }

  static async fromPrivateKey(key: Uint8Array): Promise<AIPIdentity> {
    const ed = await getEd25519();
    const pub = ed.getPublicKey(key);
    return new AIPIdentity(key, pub, publicKeyToDIDKey(pub));
  }

  get publicKeyBase64(): string {
    return typeof Buffer !== 'undefined'
      ? Buffer.from(this.publicKey).toString('base64')
      : btoa(String.fromCharCode(...this.publicKey));
  }

  async sign(data: Uint8Array): Promise<string> {
    const ed = await getEd25519();
    const sig = ed.sign(data, this.privateKey);
    return typeof Buffer !== 'undefined'
      ? Buffer.from(sig).toString('base64')
      : btoa(String.fromCharCode(...sig));
  }

  static async verify(data: Uint8Array, sigB64: string, signerDid: string): Promise<boolean> {
    const ed = await getEd25519();
    const pub = didKeyToPublicKey(signerDid);
    const sig = typeof Buffer !== 'undefined'
      ? new Uint8Array(Buffer.from(sigB64, 'base64'))
      : new Uint8Array(atob(sigB64).split('').map(c => c.charCodeAt(0)));
    try { return ed.verify(sig, data, pub); } catch { return false; }
  }

  getDIDDocument(params?: {
    services?: { type: string; endpoint: string }[];
    capabilities?: string[];
  }): DIDDocument {
    const keyId = `${this.did}#keys-1`;
    const now = new Date().toISOString();
    const doc: DIDDocument = {
      '@context': ['https://www.w3.org/ns/did/v1', 'https://w3id.org/security/suites/ed25519-2020/v1'],
      id: this.did, controller: this.did,
      verificationMethod: [{ id: keyId, type: 'Ed25519VerificationKey2020', controller: this.did, publicKeyMultibase: this.did.slice(8) }],
      authentication: [keyId], assertionMethod: [keyId],
      capabilityDelegation: [keyId], capabilityInvocation: [keyId],
      created: now, updated: now,
    };
    if (params?.services?.length) {
      doc.service = params.services.map((s, i) => ({ id: `${this.did}#service-${i}`, type: s.type, serviceEndpoint: s.endpoint }));
    }
    if (params?.capabilities?.length) doc['aip:capabilities'] = params.capabilities;
    return doc;
  }

  exportPrivateKey(): string {
    return typeof Buffer !== 'undefined'
      ? Buffer.from(this.privateKey).toString('base64')
      : btoa(String.fromCharCode(...this.privateKey));
  }
}
